index.js

```diff
@@ -0,0 +1,377 @@
+//. Right (Pair ({color: true, email: Nothing, words: []}) ([]))
+//.
+//. > parseArgs (spec) (Pair (conf) (['-e', 'a@x.org', '-e', 'b@x.org']))
+//. Right (Pair ({color: false, email: Just ('b@x.org'), words: []}) ([]))
+//.
+//. > parseArgs (spec) (Pair (conf) (['-w', 'foo', '-w', 'bar']))
+//. Right (Pair ({color: false, email: Nothing, words: ['foo', 'bar']}) ([]))
+//. ```
+//.
+//. As soon as an argument is encountered that is not a flag or an option
+//. value, that argument and any subsequent arguments are returned as the
+//. positional arguments (along with the final configuration value):
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--color', 'foo', 'bar']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) (['foo', 'bar']))
+//. ```
+//.
+//. `--` can be used to indicate that the remaining arguments are positional:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['--color', '--', '--no-color']))
+//. Right (Pair ({color: true, email: Nothing, words: []}) (['--no-color']))
+//. ```
+//.
+//. An unrecognized flag/option name results in an error message:
+//.
+//. ```javascript
+//. > parseArgs (spec) (Pair (conf) (['-x', '-y', '-z']))
+//. Left ('-x is unrecognized')
+//.
+//. > parseArgs (spec) (Pair (conf) (['--xxx', '--yyy', '--zzz']))
+//. Left ('--xxx is unrecognized')
+//. ```
+//.
+//. Here is the type of `parseArgs` with type aliases expanded:
+//.
+//. ```haskell
+//. parseArgs :: StrMap (Either (a -> a) (String -> Either String (a -> a)))
+//.           -> Pair a (Array String)
+//.           -> Either String (Pair a (Array String))
+//. ```
+//.
+//. Although names can improve readability, the following example demonstrates
+//. that this module's functionality is exposed via a single, pure function:
+//.
+//. ```javascript
+//. > parseArgs
+//. .   ({'--foo': Left (conf => ({...conf, foo: true})),
+//. .     '--bar': Right (bar => Right (conf => ({...conf, bar: Just (bar)})))})
+//. .   (Pair ({foo: false, bar: Nothing})
+//. .         (['--foo', '--bar', 'baz', 'quux']))
+//. Right (Pair ({foo: true, bar: Just ('baz')}) (['quux']))
+//. ```
+const parseArgs =
+def ('parseArgs')
+    ({})
+    ([$.StrMap (Handler (a)),
+      $.Pair (a) ($.Array ($.String)),
+      $.Either ($.String) ($.Pair (a) ($.Array ($.String)))])
+    (spec => pair => {
+       let conf = pair.fst;
+       const args = pair.snd;
+       for (let i = 0; i < args.length; i += 1) {
+         const arg = args[i];
+         if (arg === '-') {
+           return Left ('- is not a valid flag or option name');
+         } else if (arg === '--') {
+           return Right (Pair (conf) (args.slice (i + 1)));
+         } else if (arg.startsWith ('--')) {
+           if (!(Object.prototype.propertyIsEnumerable.call (spec, arg))) {
+             return Left (arg + ' is unrecognized');
+           } else if (spec[arg].isLeft) {
+             conf = spec[arg].value (conf);
```

Comment:
Instead of assigning to `conf`, could you just repeat the `return Right (Pair (...) ([]));` part? Or am I missing something? That last question should also indicate my motivation for this change request.

Reply:
Ah, I'm missing that it's happening in a for loop, and the return is outside of the for loop.